pkg/diagnostics/cluster/roles.go

```diff
@@ -0,0 +1,93 @@
+package cluster
+
+import (
+	"fmt"
+	"io/ioutil"
+
+	kerrs "k8s.io/kubernetes/pkg/api/errors"
+
+	authorizationapi "github.com/openshift/origin/pkg/authorization/api"
+	"github.com/openshift/origin/pkg/authorization/rulevalidation"
+	osclient "github.com/openshift/origin/pkg/client"
+	policycmd "github.com/openshift/origin/pkg/cmd/admin/policy"
+	"github.com/openshift/origin/pkg/diagnostics/types"
+)
+
+// ClusterRoles is a Diagnostic to check that there is a working Docker registry.
+type ClusterRoles struct {
+	ClusterRolesClient osclient.ClusterRolesInterface
+	SARClient          osclient.SubjectAccessReviews
+}
+
+const (
+	ClusterRolesName = "ClusterRoles"
+)
+
+func (d *ClusterRoles) Name() string {
+	return ClusterRolesName
+}
+
+func (d *ClusterRoles) Description() string {
+	return "Check that the ClusterRoles are up-to-date"
+}
+
+func (d *ClusterRoles) CanRun() (bool, error) {
+	if d.ClusterRolesClient == nil {
+		return false, fmt.Errorf("must have client.ClusterRolesInterface")
+	}
+	if d.SARClient == nil {
+		return false, fmt.Errorf("must have client.SubjectAccessReviews")
+	}
+
+	return userCan(d.SARClient, authorizationapi.AuthorizationAttributes{
+		Verb:     "list",
+		Resource: "clusterroles",
+	})
+}
+
+func (d *ClusterRoles) Check() types.DiagnosticResult {
+	r := types.NewDiagnosticResult(ClusterRolesName)
+
+	reconcileOptions := &policycmd.ReconcileClusterRolesOptions{
+		Confirmed:  false,
+		Union:      false,
+		Out:        ioutil.Discard,
+		RoleClient: d.ClusterRolesClient.ClusterRoles(),
+	}
+
+	changedClusterRoles, err := reconcileOptions.ChangedClusterRoles()
+	if err != nil {
+		r.Error("ClusterRolesDiagnostic1000", err, fmt.Sprintf("Error inspecting ClusterRoles: %v", err))
+	}
+
+	// success
+	if len(changedClusterRoles) == 0 {
+		return r
+	}
+
+	for _, changedClusterRole := range changedClusterRoles {
+		actualClusterRole, err := d.ClusterRolesClient.ClusterRoles().Get(changedClusterRole.Name)
+		if kerrs.IsNotFound(err) {
+			r.Error("ClusterRolesDiagnostic1002", nil, fmt.Sprintf("clusterrole/%s is missing.", changedClusterRole.Name))
+			continue
+		}
+		if err != nil {
+			r.Error("ClusterRolesDiagnostic1001", err, fmt.Sprintf("Unable to get clusterrole/%s: %v", changedClusterRole.Name, err))
+		}
+
+		_, missingRules := rulevalidation.Covers(actualClusterRole.Rules, changedClusterRole.Rules)
+		if len(missingRules) == 0 {
+			r.Warn("ClusterRolesDiagnostic1003", nil, fmt.Sprintf("clusterrole/%s has changed, but the existing role covers the new role.", changedClusterRole.Name))
+			r.Debug("ClusterRolesDiagnostic1004", fmt.Sprintf("clusterrole/%s is now %v.", changedClusterRole.Name, changedClusterRole))
+			continue
+		}
+
+		r.Error("ClusterRolesDiagnostic1005", nil, fmt.Sprintf("clusterrole/%s has changed and the existing role does not have enough permissions.", changedClusterRole.Name))
```

Comment:
Assuming you intend to give directions for remediation here... I have to admit I find the output of reconcile-cluster-roles fairly opaque; don't know if this is the right place (maybe just give a link) but any insight that can be given into the practical consequences of either ignoring this or reconciling as proposed would be appreciated.